Задачи по 6му спринту.

Danzl opened the pull request:
Добавлена работа с сетью.
В приложении использованы фильмы из топ-250 и самых популярных фильмов IMDb. Обновлён метод viewDidLoad() в MovieQuizViewController. Приложение запускается и работает корректно по функциональным и техническим требованиям. Если происходит ошибка загрузки, то показывается алерт с ошибкой. Приложение не пишет в логи сообщения об ошибках, связанных с работой с UI из другого потока.

Danzl committed:
Задачи по 6му спринту.

Добавлена работа с сетью.
В приложении использованы фильмы из топ-250 и самых популярных фильмов IMDb.
Обновлён метод viewDidLoad() в MovieQuizViewController.
Приложение запускается и работает корректно по функциональным и техническим требованиям.
Если происходит ошибка загрузки, то показывается алерт с ошибкой.
Приложение не пишет в логи сообщения об ошибках, связанных с работой с UI из другого потока.

tsaidov7 commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Models/QuizQuestion.swift:
```diff
@@ -0,0 +1,15 @@
+//
+
+import Foundation
+import UIKit
```
Необязательно импортить оба фреймворка. UIKit содержит в себе Foundation.

tsaidov7 commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Presentation/AlertPresenter.swift:
```diff
@@ -8,24 +8,19 @@
+struct AlertPresenter: AlertProtocol {
     
-    let alertModel: AlertModel
     weak var viewController: UIViewController?
```
Думаю, ссылка на контроллер должна быть приватной, раз вы ее инджектите из конструктора

tsaidov7 commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Presentation/AlertPresenter.swift:
```diff
@@ -8,24 +8,19 @@
-        let action = UIAlertAction(title: alertModel.buttonText, style: .default) { [weak self] _ in
-            self?.alertModel.completion()
-        }
+    internal func showAlert(quiz result: AlertModel) {
```
Функция и так internal. Модификатор излишен

tsaidov7 commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,44 +1,69 @@
         guard let currentQuestion = currentQuestion else { return }
-        yesButton.isEnabled = false
+        blockButtons()
         showAnswerResult(isCorrect: true == currentQuestion.correctAnswer)
```
Ни к чему сравнивать булево значение `true == true`... Вы сравниваете числа или строки (`2 == 2`; `"abc" == "cba"`), чтобы получить булево значение. Но если у вас уже есть оно, зачем сравнивать его?)
`showAnswerResult(isCorrect: currentQuestion.correctAnswer)`

tsaidov7 commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,44 +1,69 @@
         guard let currentQuestion = currentQuestion else { return }
-        noButton.isEnabled = false
+        blockButtons()
         showAnswerResult(isCorrect: false == currentQuestion.correctAnswer)
```
`showAnswerResult(isCorrect: !currentQuestion.correctAnswer)`

tsaidov7 commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,44 +1,69 @@
+    private func showLoadingIndicator() {
+        activityIndicator = UIActivityIndicatorView(frame: CGRectMake(0, 0, 40, 40))
+        self.view.addSubview(activityIndicator)
+    }
```
А для чего вам неиспользуемый метод? Не вижу его вызова. К тому же вы создали индикатор в сториборде, зачем создавать его программно?

tsaidov7 commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -7,35 +7,62 @@
+                    switch result {
+                    case .success(let mostPopularMovies):
+                        self.movies = mostPopularMovies.items
+                        self.delegate?.didLoadDateFromServer()
```
Не обработали все кейсы ошибок, которые могут прийти от сервера. Сделайте запрос https://imdb-api.com/en/API/MostPopularMovies/k_ixqybnm5аввавпавпвапав. Вы увидите 200 респонс, но при этом ошибку от сервера. Сработает данный кейс, вы получите пустой массив и отобразите пустой экран, при этом, по-хорошему, должны были отобразить алерт с ошибкой

tsaidov7 replied:
<img width="1126" alt="image" src="https://user-images.githubusercontent.com/37248555/206916000-a7e00977-037b-47cc-9de3-50d0546821e0.png">

tsaidov7 replied:
Вы здесь должны проверять пропертю errorMessage. И если она не пустая, то вызывать метод didFailToLoadData делегата

tsaidov7 commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -7,35 +7,62 @@
+    
+    func loadData() {
+        moviesLoader.loadMovies { [weak self] result in
+            DispatchQueue.main.async { [weak self] in
```
Здесь не нужно прописывать [weak self].  Вы уже слабо захватили экземпляр объекта строчкой выше

tsaidov7 commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -7,35 +7,62 @@
+                                        text: text,
+                                        correctAnswer: correctAnswer)
+            
+            DispatchQueue.main.async { [weak self] in
```
Здесь не нужно прописывать [weak self]. Вы уже слабо захватили экземпляр объекта на 40 строчке кода

Danzl replied:
Не смог понять как исправить этот момент =(

tsaidov7 commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -7,35 +7,62 @@
+            
+            DispatchQueue.main.async { [weak self] in
+                guard let self = self else { return }
+                self.delegate?.didRecieveNextQuestion(question: question)
```
Весь этот код вы должны делать в `do` блоке. Потому что иначе, когда вы словите ошибку, вы просто лог напечатаете и отобразите что? Пустой экран? Вы должны обработать ошибку
```
do {
   let imageData = try Data(contentsOf: movie.resizedImageUrl)
   let rating = Float(movie.rating) ?? 0
   let questionRating = Float.random(in: 7.50...9.00)
   let text = "Рейтинг этого фильма больше \(String(format: "%.2f", questionRating))?"
   let correctAnswer = rating > questionRating
   let question = QuizQuestion(image: imageData,
                               text: text,
                               correctAnswer: correctAnswer)

   DispatchQueue.main.async {
   ...
} catch let error {
   // Обработать ошибку!
}
```

Danzl committed:
Внесены изменения, согласно комментариям ревьюера

Убран фреймворк Foundation, там где он и так импортирован через UIKit.
Ссылка на контроллер стала приватной.
Модификатор internal убран.
Удалены сравнения одинаковых нулевых значений.
Избавился от ненужного метода showLoadingIndicator.
Обработаны дополнительные кейсы ошибок.
Необходимая часть кода вынесена в do блок.

tsaidov7 commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -13,22 +13,28 @@ class QuestionFactory: QuestionFactoryProtocol {
+                            self.movies = mostPopularMovies.items
+                            self.delegate?.didLoadDateFromServer() }
+                        else {
+                            self.delegate?.didFailToLoadData(with: errorHTTP.errorhttp)
```
А можно прям взять ошибку от сервера из поля errorMessage)
```
private enum NetworkError: Error {
   case serverError(description: String)
}

self.delegate?. didFailToLoadData(with: NetworkError.serverError(description: mostPopularMovies.errorMessage))
```

tsaidov7 commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -13,22 +13,28 @@ class QuestionFactory: QuestionFactoryProtocol {
     }
     
-    // MARK: - Func
+    private enum errorHTTP: Error {
```
Тип всегда должен начинаться с заглавной буквы

tsaidov7 commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -13,22 +13,28 @@ class QuestionFactory: QuestionFactoryProtocol {
     private let moviesLoader: MoviesLoading
     weak var delegate: QuestionFactoryDelegate?
```
private
